nova_vm/src/ecmascript/syntax_directed_operations/function_definitions.rs

```diff
@@ -198,8 +211,24 @@ pub(crate) fn evaluate_async_function_body(
         let exe = Executable::compile_function_body(agent, body, is_concise_arrow_function);
 
         match Vm::execute(agent, &exe) {
-            Ok(result) => promise_capability.resolve(agent, result.unwrap_or(Value::Undefined)),
-            Err(err) => promise_capability.reject(agent, err.value()),
+            ExecutionResult::Return(result) => promise_capability.resolve(agent, result),
+            ExecutionResult::Throw(err) => promise_capability.reject(agent, err.value()),
+            ExecutionResult::Await { vm, awaited_value } => {
+                let awaited_promise = if let Value::Promise(promise) = awaited_value {
+                    promise
+                } else {
+                    let new_capability = PromiseCapability::new(agent);
+                    new_capability.resolve(agent, awaited_value);
+                    new_capability.promise()
+                };
+                let handler = PromiseReactionHandler::Await(agent.heap.create(AwaitReaction {
+                    vm: Some(vm),
+                    executable: Some(exe),
+                    execution_context: Some(agent.running_execution_context().clone()),
```

Comment:
**question:** Is it not possible to pop it here? Though I guess the computational complexity is actually exactly equal now that I think about it.

Reply:
Well, the context is popped in `ECMAScriptFunction::internal_call` after this function returns, so popping it here as well would mess up the context stack.

Reply:
Yeah, I was wondering if we could somehow intimate to `internal_call` that we don't want it popping... But since this is actually trivially copyable data (IIRC) it shouldn't matter if we clone here and pop-to-nowhere there, or pop-to-heap here and do nothing there.